op-batcher/readme.md

```diff
@@ -0,0 +1,69 @@
+# op-batcher
+
+The `op-batcher` is responsible for ensuring data availability. See the [specs](https://specs.optimism.io/protocol/batcher.html).
+
+
+## Interactions & Dependencies
+The `op-batcher` works together with the [sequencer](../op-node/) (which it reads unsafe blocks from), the data availability layer (e.g. Layer 1 or an [Alt DA](../op-alt-da/) layer, which it posts data to), and the [derivation pipeline](../op-node/) (which reads the data from the DA layer and progresses the safe chain).
+
+It depends directly on some code shared with the derivation pipeline, namely the [`ChannelOut`](../op-node/rollup/derive/channel_out.go) implementation(s). It also depends directly on the shared [txmgr](../op-service/txmgr/) module.
+
+## Testing
+The batcher has a suite of unit test which can be triggered by running
+```
+go test ./...
+```
+from this directory. There are also end-to-end tests in [`op-e2e`](../op-e2e/) which integrate the batcher.
+
+## Architecture
+
+The architecture of this batcher implementation is shown on the left side of the following diagram:
+
+![architecture](./architecture.png)
+
+Batch submitting (writing to the DA layer, in the middle of the diagram) works together with the derivation pipeline (on the right side of the diagram, reading from the DA layer) to progress the safe chain.
+
+The philosophy behind the current architecture is:
+* Blocks, channels and frames are kept around for as long as they might be needed, and discarded as soon as they are not needed. They are not moved from one part of state to another.
+* We retain block data in a strict order for as long as necessary. We only garbage collect frames, channels and blocks when the safe head moves sufficiently and those structures have done their job.
+* When something goes wrong, we rewind the state cursors by the minimal amount we need to get going again.
+
+
+### Happy path
+
+In the happy path, the batcher periodically:
+1. Enqueues unsafe blocks and dequeues safe blocks from the sequencer to its internal state.
+2. Enqueues a new channel, if necessary.
+3. Processes some unprocessed blocks into the current channel, triggers the compression of the block data and the creation of frames.
+4. Sends frames from the channel queue to the DA layer as (e.g. to Ethereum L1 as calldata or blob transactions).
+5. If there is more transaction data to send, go to 2. Else wait for a tick and go to 1.
+
+
+The `blockCursor` state variable tracks the next unprocessed block.
+In each channel, the `frameCursor` tracks the next unsent frame.
+
+
+### Reorgs
+When an L2 unsafe reorg is detected, the batch submitter will reset its state, and wait for any in flight transactions to be ingested by the verifier nodes before starting work again.
```

Comment:
What is an L2 unsafe reorg wrt:
<img width="194" alt="image" src="https://github.com/user-attachments/assets/75f91eb3-3a48-42b6-b3b1-9f3ec70d71fc">

Even safe L2 blocks (submitted on L1) could be reorged out by an L1 reorg right? Did you mean L1 unsafe (or safe) reorg here?

Reply:
Yes safe L2 blocks can be reorged by an L1 reorg.

What I mean here by an L2 reorg, is when the unsafe blocks being pulled from the sequencer don't descend from the previous L2 blocks in the batcher's memory. Does that make sense?

Reply:
Yep that makes sense! Could we rephrase to just say "ingested by the sequencer's op-node" here? Or are there cases the batcher is really waiting for nodeS?

And just to be sure, this kind of reorg can only happen when an L1 reorg happens right..?

Reply:
Yes we can make that edit.

An unsafe reorg can happen when e.g. the batcher is down for a long time such that the sequencing window is breached https://docs.optimism.io/stack/rollup/derivation-pipeline.

Reply:
On reading the code more here, this same logic would also be applied for safe reorgs right? So think this should say both unsafe and safe reorgs?

Reply:
Yes good point. Could you send a PR for these readme changes?

Also, the logic for handling these reorgs is being refactored here #13060

Reply:
Sounds good, I'll wait for that PR to land and then refactor my PR on top of it. Also I'll include the readme changes in my PR to force you to review it :D